cargo-shuttle/src/init.rs

```diff
@@ -0,0 +1,650 @@
+            Ok(HelloWorld)
+        }"#}
+    }
+}
+
+/// Returns a framework-specific struct that implements the trait `ShuttleInit`
+/// for writing framework-specific dependencies to `Cargo.toml` and generating
+/// boilerplate code in `src/lib.rs`.
+pub fn get_framework(init_args: &InitArgs) -> Option<Box<dyn ShuttleInit>> {
+    if init_args.axum {
+        return Some(
+            Box::new(ShuttleInitAxum)
+        );
+    }
+
+    if init_args.rocket {
+        return Some(
+            Box::new(ShuttleInitRocket)
+        );
+    }
+
+    if init_args.tide {
+        return Some(
+            Box::new(ShuttleInitTide)
+        );
+    }
+
+    if init_args.tower {
+        return Some(
+            Box::new(ShuttleInitTower)
+        );
+    }
+
+    None
+}
+
+/// Interoprates with `cargo` crate and calls `cargo init --libs [path]`.
+pub fn cargo_init(path: PathBuf) -> Result<()> {
+    let opts = NewOptions::new(None, false, true, path, None, None, None)?;
+    let cargo_config = cargo::util::config::Config::default()?;
+    let init_result = cargo::ops::init(&opts, &cargo_config)?;
+
+    // Mimick `cargo init` behavior and log status or error to shell
+    cargo_config
+        .shell()
+        .status("Created", format!("{} (shuttle) package", init_result))?;
+
+    Ok(())
+}
+
+/// Processes `Cargo.toml` after calling `cargo_init` function to re-order `lib` and `dependencies`
+/// tables as well as inserting `shuttle-service` dependency to the `dependencies` table.
+pub fn process_cargo_init(path: PathBuf) -> Result<()> {
+    let cargo_toml_path = path.join("Cargo.toml");
+    let mut cargo_doc = read_to_string(cargo_toml_path.clone()).unwrap().parse::<Document>().unwrap();
+
+    // Remove empty dependencies table to re-insert after the lib table is inserted
+    cargo_doc.remove("dependencies");
+
+    // Create an empty `[lib]` table
+    cargo_doc["lib"] = Item::Table(Table::new());
+
+    // Fetch the latest shuttle-service version from crates.io
+    let manifest_path = find(Some(path.as_path())).unwrap();
+    let url = registry_url(manifest_path.as_path(), None).expect("Could not find registry URL");
+
+    // Create `[dependencies]` table
+    let mut dependencies = Table::new();
+
+    // Set "shuttle-service" version to `[dependencies]` table
+    set_inline_table_dependency_version(
+        "shuttle-service",
+        &mut dependencies,
+        &manifest_path,
+        &url,
+        get_latest_dependency_version,
+    );
+
+    cargo_doc["dependencies"] = Item::Table(dependencies);
+
+    // Truncate Cargo.toml and write the updated `Document` to it
+    let mut cargo_toml = File::create(cargo_toml_path.clone())?;
+    cargo_toml.write_all(cargo_doc.to_string().as_bytes())?;
+
+    Ok(())
+}
+
+/// Generates framework-specific dependencies to `Cargo.toml` and boilerplate code for `src/lib.rs`.
+pub fn framework_init(project_path: &PathBuf, framework: Box<dyn ShuttleInit>) -> Result<()> {
```

Comment:
Since this function and the previous one (`process_cargo_init`) both setup the `manifest_path` and `url` and both then write out a toml file, I'm wondering if it is possible to combine them to only have these steps happen once? Maybe the framework code can move into `process_cargo_init`?

This all depends on the no-op comment...

Reply:
So if the no-op part stays, I can ignore this comment? :D

Reply:
Lol, I think I was wrong about an earlier request to remove the `Default` enum variant :sweat_smile: 

If we create a `ShuttleInitNoOp` and return it instead of `None` in `get_framework`, then we should be able to make `framework: Box<dyn ShuttleInit>` an argument of `process_cargo_init` right?